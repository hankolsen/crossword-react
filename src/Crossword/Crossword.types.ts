export type CellType = {
  [direction in Direction]: string[];
} & {
  column: number;
  number: number;
  row: number;
  highlighted?: boolean;
  selected?: boolean;
  text?: string;
};

export enum Direction {
  across = 'across',
  down = 'down',
}

export type Selection = {};

export type Position = { x: number; y: number };

export type SeparatorType = {
  direction: Direction;
  position: Position;
  separator: string;
  locations: number[];
  id: string;
};
