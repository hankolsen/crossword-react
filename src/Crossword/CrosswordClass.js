/* eslint-disable no-param-reassign */
import React from 'react';
import PropTypes from 'prop-types';

import './Crossword.css';
import { Redirect } from 'react-router-dom';
import Separators from '../Separators/Separators';
import Cells from '../Cells/Cells';
import CellInput from '../Cells/CellInput/CellInput';
import LoadingIndicator from '../LoadingIndicator/LoadingIndicator';
import {
  cellContainsOtherDirection,
  createCrossword,
  cellIsStartingWord,
  deselectAll,
  emptyAll,
  getCurrentId,
  getInputPosition,
  highlightId,
  isIgnorableKey,
  isValidKey,
  toggleDirection,
} from './crosswordHelper';

class Crossword extends React.Component {
  constructor(props) {
    super(props);

    this.separators = [];
    this.numberOfColumns = 0;
    this.numberOfRows = 0;
    this.inputWidth = 0;
    this.inputHeight = 0;
    this.boardWidth = 0;
    this.boardHeight = 0;

    this.state = {
      isLoading: true,
      direction: 'across',
      error: false,
      showInput: false,
    };

    this.clickHandler = this.clickHandler.bind(this);
    this.inputHandler = this.inputHandler.bind(this);
    this.keyUpHandler = this.keyUpHandler.bind(this);
    this.inputClickHandler = this.inputClickHandler.bind(this);
    this.reset = this.reset.bind(this);
  }

  componentDidMount() {
    const { match } = this.props;
    this.crosswordId = match.params.id;
    createCrossword(this.crosswordId)
      .then(
        ({
          name,
          cells,
          separators,
          numberOfColumns,
          numberOfRows,
          boardWidth,
          boardHeight,
          inputWidth,
          inputHeight,
        }) => {
          Object.assign(this, {
            name,
            separators,
            numberOfColumns,
            numberOfRows,
            boardWidth,
            boardHeight,
            inputWidth,
            inputHeight,
          });

          this.setState({ cells });
        },
      )
      .catch(() => {
        this.setState({ error: true });
      })
      .finally(() => this.setState({ isLoading: false }));
  }

  handleArrowMove({ arrow }) {
    const { cells, currentCell } = this.state;
    const { row, column } = currentCell;

    /* eslint-disable no-shadow */
    const moves = {
      Left: (cells, row, column) => cells[row][column - 1],
      Right: (cells, row, column) => cells[row][column + 1],
      Up: (cells, row, column) => cells[row - 1] && cells[row - 1][column],
      Down: (cells, row, column) => cells[row + 1] && cells[row + 1][column],
    };
    /* eslint-enable no-shadow */

    const newCell = moves[arrow](cells, row, column);
    if (newCell) {
      this.moveTo(newCell);
    }
  }

  inputClickHandler() {
    let { direction, currentCell } = this.state;
    const { cells, selection } = this.state;

    if (currentCell) {
      if (cellContainsOtherDirection({ currentCell, direction })) {
        direction = toggleDirection(direction);
        this.highlightCurrentSelection({ direction });
      } else if (currentCell[direction].length > 1) {
        const currentIndex = currentCell[direction].indexOf(selection);
        const nextIndex = (currentIndex + 1) % currentCell[direction].length;
        const newId = currentCell[direction][nextIndex];
        highlightId({ cells, direction, id: newId, currentCell });
        this.setState({ cells, selection: newId });
      }
    } else {
      currentCell = cells[0][0]; // eslint-disable-line prefer-destructuring
      if (!currentCell[direction]) {
        direction = toggleDirection(direction);
      }
      this.setState({ currentCell, direction }, () => {
        this.highlightCurrentSelection({ direction });
      });
    }
  }

  clickHandler(event, row, column) {
    let { direction } = this.state;
    const { cells } = this.state;
    const currentCell = cells[row][column];
    if (
      currentCell.number &&
      !currentCell.highlighted &&
      !cellIsStartingWord({ cell: currentCell, direction }) &&
      cellContainsOtherDirection({ currentCell, direction }) &&
      cellIsStartingWord({
        cell: currentCell,
        direction: toggleDirection(direction),
      })
    ) {
      direction = toggleDirection(direction);
    }
    if (!currentCell[direction] || currentCell.selected) {
      direction = toggleDirection(direction);
    }
    this.setState({ showInput: true });
    this.setState({ currentCell }, () => {
      this.highlightCurrentSelection({ direction }).then(
        ({ direction: newDirection }) => {
          this.setState({ cells, direction: newDirection, currentCell });
        },
      );
    });
  }

  highlightCurrentSelection({ direction }) {
    const { cells, currentCell, selection } = this.state;
    return new Promise((resolve) => {
      const id = getCurrentId({ currentCell, direction, selection });
      if (id) {
        highlightId({ cells, direction, id, currentCell });
      }
      this.setState({ direction, selection: id }, () => resolve({ direction }));
    });
  }

  reset() {
    localStorage.removeItem(`kryzz-${this.crosswordId}`);
    const { cells } = this.state;
    emptyAll(cells);
    this.setState({ cells });
  }

  inputHandler(event) {
    const { currentCell } = this.state;
    const key = event.target.value;
    if (isValidKey(key)) {
      currentCell.text = key.toUpperCase();
      this.setState({ currentCell });
      this.moveToNext();
    }
  }

  keyUpHandler(event) {
    const { key } = event;

    if (key === 'Shift' || event.metaKey) {
      return;
    }

    // Is it an arrow key?
    const [, arrow] = key.match(/Arrow(\w+)$/) || [];

    if (arrow) {
      this.handleArrowMove({ arrow });
    }

    if (isIgnorableKey(key)) {
      event.preventDefault();
      return;
    }

    if (key === 'Backspace') {
      const { currentCell } = this.state;
      if (currentCell.text) {
        currentCell.text = '';
        this.setState({ currentCell });
      } else {
        this.moveToPrevious();
      }
    }

    const { cells } = this.state;
    const entries = cells.map((row) => row.map((cell) => cell && cell.text));
    localStorage.setItem(`kryzz-${this.crosswordId}`, JSON.stringify(entries));
  }

  moveTo(nextCell) {
    const { cells } = this.state;
    let { direction, selection } = this.state;
    if (!nextCell[direction] || !nextCell[direction].includes(selection)) {
      direction = toggleDirection(direction);
      if (!nextCell[direction] || !nextCell[direction].length) {
        direction = toggleDirection(direction);
      }
      [selection] = nextCell[direction];
    }
    highlightId({ cells, direction, id: selection, currentCell: nextCell });
    this.setState({ cells, selection, currentCell: nextCell, direction });
  }

  moveToPrevious() {
    this.moveToNext(-1);
  }

  moveToNext(dir = 1) {
    const { direction, cells, selection } = this.state;
    const array = [];

    cells.map((row) =>
      row.forEach((cell) => {
        if (cell && cell[direction] && cell[direction].includes(selection)) {
          array.push(cell);
        }
      }),
    );

    let nextCell;
    const currentIndex = array.findIndex((cell) => cell.selected);
    if (currentIndex > -1 && currentIndex < array.length) {
      const nextIndex = currentIndex + dir;
      nextCell = array[nextIndex];
    }

    if (nextCell) {
      deselectAll(cells);
      nextCell.selected = true;
      this.setState({ currentCell: nextCell, cells });
    }
  }

  renderCrossword() {
    const { cells, currentCell = {}, showInput } = this.state;
    const { left, top } = getInputPosition(currentCell);
    return (
      <div className="crossword-board">
        <svg
          className="crossword__grid"
          viewBox={`0 0 ${this.boardWidth} ${this.boardHeight}`}
          fill="#222222"
        >
          <rect
            x="0"
            y="0"
            width={this.boardWidth}
            height={this.boardHeight}
            className="crossword__grid-background"
          />
          {cells && <Cells cells={cells} clickHandler={this.clickHandler} />}
          <Separators separators={this.separators} />
        </svg>
        {showInput && (
          <CellInput
            top={top}
            left={left}
            width={this.inputWidth}
            height={this.inputHeight}
            inputHandler={this.inputHandler}
            keyUpHandler={this.keyUpHandler}
            clickHandler={this.inputClickHandler}
          />
        )}
      </div>
    );
  }

  render() {
    const { isLoading, error } = this.state;

    if (error) {
      return <Redirect to="/" />;
    }

    return (
      <div className="crossword">
        <h3>{this.name ? this.name : 'Loading crossword'}</h3>
        <div className="crossword-container">
          {isLoading ? <LoadingIndicator /> : this.renderCrossword()}
        </div>
        <button
          onClick={this.reset}
          type="submit"
          className={isLoading ? 'hidden' : ''}
        >
          Reset
        </button>
      </div>
    );
  }
}

Crossword.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string,
    }).isRequired,
  }).isRequired,
};

export default Crossword;
