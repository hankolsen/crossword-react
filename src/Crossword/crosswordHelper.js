import { CELL_HEIGHT, CELL_WIDTH } from 'config';

let boardHeight;
let boardWidth;
let numberOfRows;
let numberOfColumns;

const dehighlightAll = (cells) => {
  cells.forEach((row) =>
    row.forEach((cell) => {
      if (cell) {
        cell.highlighted = false;
      }
    }),
  );
};

const deselectAll = (cells) => {
  if (!cells) {
    return;
  }
  cells.forEach((row) =>
    row.forEach((cell) => {
      if (cell) {
        cell.selected = false;
      }
    }),
  );
};

const emptyAll = (cells) => {
  cells.map((row) =>
    row.forEach((column) => {
      if (column) {
        column.text = '';
      }
    }),
  );
};

/* eslint-disable no-confusing-arrow */
const toggleDirection = (direction) =>
  direction === 'across' ? 'down' : 'across';
/* eslint-enable no-confusing-arrow */

const fillCell = ({
  cells,
  row,
  column,
  index,
  number,
  id,
  direction,
  text,
  hasTurn,
}) => {
  cells[row][column] = {
    ...cells[row][column],
    text,
    number:
      index === 0 ? number : cells[row][column] && cells[row][column].number,
    row,
    column,
  };

  if (cells[row][column] && cells[row][column][direction]) {
    if (hasTurn && index >= hasTurn[0]) {
      cells[row][column][direction] = [...cells[row][column][direction], id];
    } else {
      cells[row][column][direction] = [id, ...cells[row][column][direction]];
    }
  } else {
    cells[row][column][direction] = [id];
  }
};

const createCrossword = (crosswordId) =>
  new Promise((resolve, reject) =>
    fetch(`/.netlify/functions/crosswords/${crosswordId}`)
      .then((response) => response.json())
      .then(({ crossword }) => {
        if (!crossword) {
          reject();
        }
        numberOfColumns = crossword.size.width;
        numberOfRows = crossword.size.height;
        const separators = [];

        const cells = Array(numberOfRows)
          .fill()
          .map(() => Array(numberOfColumns).fill());

        const userData = JSON.parse(
          localStorage.getItem(`kryzz-${crosswordId}`) || 'null',
        );

        crossword.entries.sort(
          (a, b) => a.position.y - b.position.y || a.position.x - b.position.x,
        );

        let number = 1;
        const numbers = {};
        crossword.entries.forEach(
          ({ direction, position, length, separatorLocations, turns }) => {
            let column = position.x;
            let row = position.y;
            let walkingDirection = direction;
            let turnIndex = 0;
            if (numbers[`${column},${row}`]) {
              number = numbers[`${column},${row}`];
            }
            const id = `${number}-${direction}`;
            numbers[`${column},${row}`] = number;
            Array(length)
              .fill()
              .forEach((_, i) => {
                const text = userData && userData[row][column];
                fillCell({
                  cells,
                  row,
                  column,
                  index: i,
                  number,
                  id,
                  direction,
                  text,
                  hasTurn: turns,
                });

                if (turns && turns.length && turns[turnIndex] - 1 === i) {
                  walkingDirection = toggleDirection(walkingDirection);
                  cells[row][column].arrow = walkingDirection;
                  turnIndex += 1;
                }

                if (walkingDirection === 'across') {
                  column += 1;
                } else {
                  row += 1;
                }
              });

            Object.entries(separatorLocations || {}).forEach(
              ([separator, locations]) => {
                if (locations && locations.length) {
                  separators.push({
                    direction,
                    position,
                    separator,
                    locations,
                    id,
                  });
                }
              },
            );
            number += 1;
          },
        );

        const inputWidth = 100 / numberOfColumns;
        const inputHeight = 100 / numberOfRows;
        boardWidth = CELL_WIDTH * numberOfColumns + numberOfColumns + 1 || 0;
        boardHeight = CELL_HEIGHT * numberOfRows + numberOfRows + 1 || 0;
        const { name } = crossword;
        // eslint-disable-next-line max-len
        resolve({
          name,
          cells,
          separators,
          numberOfColumns,
          numberOfRows,
          boardWidth,
          boardHeight,
          inputWidth,
          inputHeight,
        });
      })
      .catch((e) => reject(e)),
  );

const isValidKey = (key) => key.match(/^[a-zåäö]{1}$/i) || key === 'Backspace';

const isIgnorableKey = (key) => key === 'Tab' || !isValidKey(key);

const getCurrentId = ({ currentCell, direction, selection }) => {
  let id;
  if (
    selection &&
    currentCell &&
    currentCell[direction] &&
    currentCell[direction].includes(selection)
  ) {
    id = selection;
  } else {
    id = currentCell[direction] && currentCell[direction][0];
  }
  /* if (!id) {
    direction = toggleDirection(direction);
    [id] = currentCell[direction];
  } */
  return id;
};

const highlightId = ({ cells, direction, id, currentCell }) => {
  deselectAll(cells);
  dehighlightAll(cells);
  currentCell.selected = true;

  cells.map((row) =>
    row.forEach((cell) => {
      if (cell && cell[direction] && cell[direction].includes(id)) {
        cell.highlighted = true;
      }
    }),
  );
};

const cellIsStartingWord = ({ cell, direction }) =>
  cell &&
  cell[direction] &&
  cell[direction].find((id) => id.startsWith(cell.number));

const cellIsStartingWordInOtherDirection = ({ cell, direction }) => {
  const otherDirection = direction === 'across' ? 'down' : 'across';
  return cellIsStartingWord({ cell, direction: otherDirection });
};

const cellContainsOtherDirection = ({ currentCell, direction }) =>
  (direction === 'across' && currentCell.down) ||
  (direction === 'down' && currentCell.across);

export {
  createCrossword,
  cellContainsOtherDirection,
  cellIsStartingWord,
  cellIsStartingWordInOtherDirection,
  emptyAll,
  getCurrentId,
  highlightId,
  isIgnorableKey,
  isValidKey,
  toggleDirection,
};
