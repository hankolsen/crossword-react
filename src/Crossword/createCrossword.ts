import { CELL_HEIGHT, CELL_WIDTH } from '../config';
import { CellType } from './Crossword.types';
import { toggleDirection } from './crosswordHelper';
import fillCell from './fillCell';

type Entry = {
  id: string;
  direction: 'across' | 'down';
  group: string[];
  length: number;
  position: { x: number; y: number };
  separatorLocations?: { [key: string]: number[] };
  turns?: number[];
};

export type CrosswordType = {
  name: string;
  cells: CellType[][];
  separators?: unknown[];
  numberOfColumns: number;
  numberOfRows: number;
  boardWidth: number;
  boardHeight: number;
  inputWidth: number;
  inputHeight: number;
};

type CrosswordResponse = {
  crossword: {
    entries: Entry[];
    name: string;
    size: { width: number; height: number };
    _id: string;
  };
};

const createCrossword = (crosswordId: string): Promise<CrosswordType> =>
  new Promise((resolve, reject) =>
    fetch(`/.netlify/functions/crosswords/${crosswordId}`)
      .then((response) => response.json())
      .then(({ crossword }: CrosswordResponse) => {
        if (!crossword) {
          reject();
        }
        const numberOfColumns = crossword.size.width;
        const numberOfRows = crossword.size.height;
        const separators: unknown[] = [];

        const cells = Array(numberOfRows)
          .fill(null)
          .map(() => Array(numberOfColumns).fill(null));

        const userData = JSON.parse(
          localStorage.getItem(`kryzz-${crosswordId}`) || 'null',
        );

        crossword.entries.sort(
          (a, b) => a.position.y - b.position.y || a.position.x - b.position.x,
        );

        let number = 1;
        const numbers: { [key: string]: number } = {};
        crossword.entries.forEach(
          ({ direction, position, length, separatorLocations, turns }) => {
            let column = position.x;
            let row = position.y;
            let walkingDirection = direction;
            let turnIndex = 0;
            if (numbers[`${column},${row}`]) {
              number = numbers[`${column},${row}`];
            }
            const id = `${number}-${direction}`;
            numbers[`${column},${row}`] = number;
            Array(length)
              .fill(null)
              .forEach((_, i) => {
                const text = userData && userData[row][column];
                fillCell({
                  cells,
                  row,
                  column,
                  index: i,
                  number,
                  id,
                  direction,
                  text,
                  hasTurn: turns,
                });

                if (turns && turns.length && turns[turnIndex] - 1 === i) {
                  walkingDirection = toggleDirection(walkingDirection);
                  cells[row][column].arrow = walkingDirection;
                  turnIndex += 1;
                }

                if (walkingDirection === 'across') {
                  column += 1;
                } else {
                  row += 1;
                }
              });

            Object.entries(separatorLocations || {}).forEach(
              ([separator, locations]) => {
                if (locations && locations.length) {
                  separators.push({
                    direction,
                    position,
                    separator,
                    locations,
                    id,
                  });
                }
              },
            );
            number += 1;
          },
        );

        const inputWidth = 100 / numberOfColumns;
        const inputHeight = 100 / numberOfRows;
        const boardWidth =
          CELL_WIDTH * numberOfColumns + numberOfColumns + 1 || 0;
        const boardHeight = CELL_HEIGHT * numberOfRows + numberOfRows + 1 || 0;
        const { name } = crossword;
        console.log('cells', cells);
        // eslint-disable-next-line max-len
        resolve({
          name,
          cells,
          separators,
          numberOfColumns,
          numberOfRows,
          boardWidth,
          boardHeight,
          inputWidth,
          inputHeight,
        });
      })
      .catch((e) => reject(e)),
  );

export default createCrossword;
