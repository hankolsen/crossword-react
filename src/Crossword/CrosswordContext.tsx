/* eslint-disable @typescript-eslint/no-unused-vars */
import React, {
  createContext,
  FunctionComponent,
  useContext,
  useEffect,
  useState,
} from 'react';
import { useParams } from 'react-router-dom';
import createCrossword, { CrosswordType } from './createCrossword';

type CrosswordContextType = {
  clickHandler: (row: number, column: number) => void;
  crossword?: CrosswordType;
  currentCell: { row: number; column: number };
  isLoading: boolean;
  inputClickHandler: () => void;
  inputHandler: (event: React.KeyboardEvent<HTMLInputElement>) => void;
  keyUpHandler: (event: React.KeyboardEvent<HTMLInputElement>) => void;
  showInput: boolean;
};

const CrosswordContext = createContext<CrosswordContextType | undefined>(
  undefined,
);

const useCrossword = () => {
  const context = useContext(CrosswordContext);
  if (!context) {
    throw new Error('useCrossword must be used within a CrosswordContext');
  }
  return context;
};

const CrosswordProvider: FunctionComponent = ({ children }) => {
  const [crossword, setCrossword] = useState<CrosswordType>();
  const [isLoading, setIsLoading] = useState(true);
  const [showInput, setShowInput] = useState(false);
  const [currentCell, setCurrentCell] = useState({ row: 0, column: 0 });
  const { id } = useParams<{ id: string }>();

  const clickHandler = (row: number, column: number) => {
    if (crossword) {
      setCurrentCell(crossword.cells[row][column]);
    }
    setShowInput(true);
  };

  const inputHandler = (event: React.KeyboardEvent<HTMLInputElement>) => {};
  const keyUpHandler = (event: React.KeyboardEvent<HTMLInputElement>) => {};
  const inputClickHandler = () => {};

  useEffect(() => {
    setIsLoading(true);
    createCrossword(id)
      .then((crossword) => setCrossword(crossword))
      .catch()
      .finally(() => setIsLoading(false));
  }, [id]);

  const value = {
    clickHandler,
    crossword,
    currentCell,
    isLoading,
    inputClickHandler,
    inputHandler,
    keyUpHandler,
    showInput,
  };

  return (
    <CrosswordContext.Provider value={value}>
      {children}
    </CrosswordContext.Provider>
  );
};

export { CrosswordProvider, useCrossword };
