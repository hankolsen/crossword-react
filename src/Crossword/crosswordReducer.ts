import { CellType, Direction } from './Crossword.types';
import { CrosswordType } from './utils/createCrossword';
import deSelectAll from './utils/deSelectAll';
import emptyAll from './utils/emptyAll';
import highlightCurrentSelection from './utils/highlightCurrentSelection';
import isArrowKey from './utils/isArrowKey';
import isIgnorableKey from './utils/isIgnorableKey';
import isValidKey from './utils/isValidKey';
import getCurrentId from './utils/getCurrentId';
import cellIsStartingWord from './utils/cellIsStartingWord';
import { moveToNext } from './utils/move';
import toggleDirection from './utils/toggleDirection';
import cellContainsOtherDirection from './utils/cellContainsOtherDirection';
import cellIsStartingWordInOtherDirection from './utils/cellIsStartingWordInOtherDirection';

type CrosswordState = {
  crossword?: CrosswordType;
  isLoading: boolean;
  showInput: boolean;
  currentCell: CellType;
  selection: string;
  direction: Direction;
};

type CrosswordAction =
  | { type: 'click_cell'; row: number; column: number }
  | { type: 'click_input' }
  | { type: 'on_input'; event: React.KeyboardEvent<HTMLInputElement> }
  | { type: 'key_up'; event: React.KeyboardEvent<HTMLInputElement> }
  | { type: 'reset' }
  | { type: 'crossword_loaded'; crossword: CrosswordType }
  | { type: 'crossword_load_failed'; err: any }
  | { type: 'load' };

const crosswordReducer = (
  state: CrosswordState,
  action: CrosswordAction,
): CrosswordState => {
  const { crossword, selection } = state;
  const { cells = [[]], crosswordId } = crossword ?? {};
  let { direction, currentCell } = state;
  switch (action.type) {
    case 'reset': {
      if (crosswordId !== undefined) {
        localStorage.removeItem(`kryzz-${crosswordId}`);
      }
      emptyAll(cells);
      return { ...state, crossword };
    }
    case 'key_up': {
      const { event } = action;
      const { key, metaKey } = event;
      if (isIgnorableKey(key) || metaKey) {
        event.preventDefault();
        return state;
      }

      if (isArrowKey(key)) {
        return state;
      }

      return state;
    }
    case 'on_input': {
      const { value } = action.event.target as HTMLInputElement;
      if (!isValidKey(value)) {
        return state;
      }

      currentCell.text = value.toUpperCase();
      const nextCell = moveToNext({
        cells,
        direction,
        selection,
        currentCell,
      });
      deSelectAll(cells);
      nextCell.selected = true;
      return {
        ...state,
        crossword,
        currentCell: nextCell,
      };
    }
    case 'click_input': {
      if (currentCell) {
        if (cellContainsOtherDirection({ currentCell, direction })) {
          direction = toggleDirection(direction);
          const id = getCurrentId({ currentCell, direction, selection });
          highlightCurrentSelection({ cells, currentCell, direction, id });
          return { ...state, selection: id, direction };
        }

        if (currentCell[direction].length > 1) {
          const currentIndex = currentCell[direction].indexOf(selection);
          const nextIndex = (currentIndex + 1) % currentCell[direction].length;
          const newId = currentCell[direction][nextIndex];
          highlightCurrentSelection({
            cells,
            currentCell,
            direction,
            id: newId,
          });
          return {
            ...state,
            selection: newId,
          };
        }
      } else {
        // eslint-disable-next-line prefer-destructuring
        currentCell = cells[0][0];
        if (!currentCell[direction]) {
          direction = toggleDirection(direction);
        }
        const id = getCurrentId({ currentCell, direction, selection });
        highlightCurrentSelection({ cells, currentCell, direction, id });
        return {
          ...state,
          direction,
          selection: id,
          currentCell,
        };
      }
      return state;
    }
    case 'click_cell': {
      const { row, column } = action;
      const currentCell = cells[row][column];

      if (
        currentCell.number &&
        !currentCell.highlighted &&
        !cellIsStartingWord({ cell: currentCell, direction }) &&
        cellContainsOtherDirection({ currentCell, direction }) &&
        cellIsStartingWordInOtherDirection({
          cell: currentCell,
          direction,
        })
      ) {
        direction = toggleDirection(direction);
      }

      if (!currentCell[direction] || currentCell.selected) {
        direction = toggleDirection(direction);
      }

      const id = getCurrentId({
        currentCell,
        direction,
        selection: state.selection,
      });

      highlightCurrentSelection({ cells, currentCell, direction, id });

      return {
        ...state,
        selection: id,
        direction,
        showInput: true,
        currentCell,
        crossword,
      };
    }
    case 'load':
      return { ...state, isLoading: true };
    case 'crossword_loaded':
      return {
        ...state,
        isLoading: false,
        crossword: action.crossword,
      };
    case 'crossword_load_failed':
      return {
        ...state,
        isLoading: false,
        crossword: undefined,
      };
    default:
      return state;
  }
};

export default crosswordReducer;
