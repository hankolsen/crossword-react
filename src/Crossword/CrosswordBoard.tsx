import React from 'react';
import CellInput from 'Cells/CellInput/CellInput';
import Cells from 'Cells/Cells';
import Separators from 'Separators/Separators';
import { useCrossword } from './CrosswordContext';

const CrosswordBoard = () => {
  const { boardWidth, boardHeight, separators } = useCrossword();
  return (
    <div className="crossword-board">
      <svg
        className="crossword__grid"
        viewBox={`0 0 ${boardWidth} ${boardHeight}`}
        fill="#222222"
      >
        <rect
          x="0"
          y="0"
          width={boardWidth}
          height={boardHeight}
          className="crossword__grid-background"
        />
        <Cells />
        <Separators separators={separators} />
      </svg>
      <CellInput />
    </div>
  );
};

export default CrosswordBoard;
