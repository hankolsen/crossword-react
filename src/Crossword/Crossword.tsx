import React from 'react';
import LoadingIndicator from '../LoadingIndicator/LoadingIndicator';
import CrosswordBoard from './CrosswordBoard';
import { useCrossword } from './CrosswordContext';

import './Crossword.css';
import ResetButton from './ResetButton';

const Crossword = () => {
  const { isLoading, name } = useCrossword();

  if (isLoading) {
    return (
      <div className="crossword">
        <h3>Loading crossword</h3>
        <LoadingIndicator />
      </div>
    );
  }

  return (
    <div className="crossword">
      <h3>{name}</h3>
      <div className="crossword-container">
        <CrosswordBoard />
      </div>
      <ResetButton />
    </div>
  );
};

export default Crossword;
