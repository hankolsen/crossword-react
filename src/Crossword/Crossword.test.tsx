import React from 'react';
import { render, screen } from 'test/test-utils';
import Crossword from './Crossword';
import { CellType } from './Crossword.types';
import { CrosswordContext, CrosswordContextType } from './CrosswordContext';

describe('Crossword test', () => {
  it('should show a loading message and spinner when fetching the crossword', () => {
    const { container } = render(
      <CrosswordContext.Provider value={{ isLoading: true } as any}>
        <Crossword />
      </CrosswordContext.Provider>,
    );
    screen.getByText(/loading crossword/i);
    expect(container).toMatchSnapshot();
  });

  it('should show a crossword when fetching is done', () => {
    const value: CrosswordContextType = {
      cells: [[]],
      boardHeight: 100,
      boardWidth: 100,
      clickHandler: () => {},
      crosswordId: 'abc-123',
      currentCell: {} as CellType,
      inputClickHandler: () => {},
      inputWidth: 10,
      inputHeight: 10,
      inputHandler: () => {},
      isLoading: false,
      reset: () => {},
      showInput: false,
      keyUpHandler: () => {},
      name: 'Tema Covers',
    };

    const { container } = render(
      <CrosswordContext.Provider value={value}>
        <Crossword />
      </CrosswordContext.Provider>,
    );
    screen.getByText(/tema covers/i);
    expect(container).toMatchSnapshot();
  });
});
