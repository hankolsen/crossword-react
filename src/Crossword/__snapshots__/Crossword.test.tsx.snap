// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Crossword test should show a crossword when fetching is done 1`] = `
<div>
  <div>
    <div
      class="crossword"
    >
      <h3>
        Tema Covers
      </h3>
      <div
        class="crossword-container"
      >
        <div
          class="crossword-board"
        >
          <svg
            class="crossword__grid"
            fill="#222222"
            viewBox="0 0 100 100"
          >
            <rect
              class="crossword__grid-background"
              height="100"
              width="100"
              x="0"
              y="0"
            />
          </svg>
        </div>
      </div>
      <button
        class=""
        type="submit"
      >
        Reset
      </button>
    </div>
  </div>
</div>
`;

exports[`Crossword test should show a loading message and spinner when fetching the crossword 1`] = `
<div>
  <div>
    <div
      class="crossword"
    >
      <h3>
        Loading crossword
      </h3>
      <svg
        class="loadingindicator__wrapper"
        height="360"
        version="1.1"
        viewBox="0 0 360 360"
        width="360"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          transform="translate(0)"
        >
          <rect
            id="background"
            x="0"
            y="0"
          />
          <rect
            class="square"
            id="rect-11"
            x="8"
            y="8"
          />
          <rect
            class="square"
            id="rect-12"
            x="96"
            y="8"
          />
          <rect
            class="square"
            id="rect-13"
            x="184"
            y="8"
          />
          <rect
            class="square"
            id="rect-14"
            x="272"
            y="8"
          />
          <rect
            class="square"
            id="rect-21"
            x="8"
            y="96"
          />
          <rect
            class="square"
            id="rect-22"
            x="96"
            y="96"
          />
          <rect
            class="square square__inverse"
            id="rect-23"
            x="184"
            y="96"
          />
          <rect
            class="square"
            id="rect-24"
            x="272"
            y="96"
          />
          <rect
            class="square"
            id="rect-31"
            x="8"
            y="184"
          />
          <rect
            class="square square__inverse"
            id="rect-32"
            x="96"
            y="184"
          />
          <rect
            class="square"
            id="rect-33"
            x="184"
            y="184"
          />
          <rect
            class="square"
            id="rect-34"
            x="272"
            y="184"
          />
          <rect
            class="square"
            id="rect-41"
            x="8"
            y="272"
          />
          <rect
            class="square"
            id="rect-42"
            x="96"
            y="272"
          />
          <rect
            class="square square__inverse"
            id="rect-43"
            x="184"
            y="272"
          />
          <rect
            class="square"
            id="rect-44"
            x="272"
            y="272"
          />
        </g>
      </svg>
    </div>
  </div>
</div>
`;
